api/auth_token.go

```diff
@@ -15,13 +15,17 @@ func (a *Auth) Token() *TokenAuth {
 }
 
 func (c *TokenAuth) Create(opts *TokenCreateRequest) (*Secret, error) {
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	defer cancelFunc()
+	return c.CreateWithContext(ctx, opts)
+}
+
+func (c *TokenAuth) CreateWithContext(ctx context.Context, opts *TokenCreateRequest) (*Secret, error) {
 	r := c.c.NewRequest("POST", "/v1/auth/token/create")
 	if err := r.SetJSONBody(opts); err != nil {
 		return nil, err
 	}
 
-	ctx, cancelFunc := context.WithCancel(context.Background())
-	defer cancelFunc()
 	resp, err := c.c.RawRequestWithContext(ctx, r)
```

Comment:
I believe the pattern of calling `context.WithCancel`  was a bit of a hack to prevent context leaks introduced in `RawRequestWithContext`:

https://github.com/hashicorp/vault/blob/8b1e041c61ba7ec61c85584fa5fcaa6f6cc0704f/api/client.go#L1160-L1166

`RawRequestWithContext()` currently requires a cancellable context, which our current code does by passing it `context.WithCancel(context.Background())`. Unfortunately, moving these lines into the wrapper will make `CreateWithContext` (and other `__WithContext` methods) prone to context leaks:

```go
// client code: OK
ctx, cancel := context.WithTimeout(context.Background(), 100*time.Millisecond)
defer cancel()
c.CreateWithContext(ctx, opts)
```
```go
// client code: POSSIBLE CONTEXT LEAK
c.CreateWithContext(context.Background(), opts)
```
In the second example, if a user has a timeout set at the client level but the timeout is never hit, the context will never be cancelled and the resources will be leaked.

One way I can think of to solve this issue is to move the timeout logic out of `RawRequestWithContext` and into each individual `___WithContext`  function:
```go
func (c *TokenAuth) Create(opts *TokenCreateRequest) (*Secret, error) {
	return c.CreateWithContext(context.Background(), opts)
}

func (c *TokenAuth) CreateWithContext(ctx context.Context, opts *TokenCreateRequest) (*Secret, error) {
    // ...
    
    if c.c.config.Timeout != 0 {
        var cancelFunc context.CancelFunc
        ctx, cancelFunc = context.WithTimeout(context.Background(), c.c.config.Timeout)
        defer cancelFunc()
    }

    resp, err := c.c.rawRequestWithContext(ctx, r)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

    // ...
}
```

Additionally, since there are some [non-api callers](https://github.com/hashicorp/vault/blob/7e74bebea84b098d9d2de0eed2490f352481717a/vault/raft.go#L1135) of `RawRequestWithContext`, it should be refactored as well:

```go
func (c* Client) RawRequestWithContext(ctx context.Context, r *Request) (*Response, error) {
     	if c.config.Timeout != 0 {
		// Note: we purposefully do not call cancel manually. The reason is
		// when canceled, the request.Body will EOF when reading due to the way
		// it streams data in. Cancel will still be run when the timeout is
		// hit, so this doesn't really harm anything.
		ctx, _ = context.WithTimeout(ctx, c.config.Timeout)
	}
	
	return rawRequestWithContext(ctx, r)
}

func (c* Client) rawRequestWithContext(ctx context.Context, r *Request) (*Response, error) {
    // the rest of the RawRequestWithContext implementation
}
```

Admittedly, this adds a few more lines to each `_WithContext` function and changes may not be an ideal solution. However, it ensures that:

1. the context is only cancelled after resp.Body.Close()
2. `_WithContext()` functions can be safely called with `context.Background()` without any potential context leaks

I’d be happy to discuss any other potential fixes to the context leak issue or help with the implementation if needed. Please let me know what you think :slightly_smiling_face:

Reply:
Good idea. I suppose to add the function like this

```go
func (c *Client) defaultContext() (context.Context, context.CancelFunc) {
	if c.config.Timeout > 0 {
		return context.WithTimeout(context.Background(), c.config.Timeout)
	}
	return context.WithCancel(context.Background())
}
```

And use it to obtain context

```go
func (c *TokenAuth) Create(opts *TokenCreateRequest) (*Secret, error) {
	ctx, cancelFunc := c.c.defaultContext()
	defer cancelFunc()
	return c.CreateWithContext(ctx, opts)
}
```

What do you think?

Reply:
I don't think this would solve the issue, unfortunately.

I apologize if my explanation was not very clear. The `Create` function codepath is actually OK with your current implementation. The issue is with `CreateWithContext` function (as well as all the other `_WithContext` functions. Since the following two lines are removed:

```diff
-	ctx, cancelFunc := context.WithCancel(context.Background())
-	defer cancelFunc()
```

... the following code will have a context leak:

```go
	config := vault.DefaultConfig()
	config.Timeout = 30 * time.Second  // set the request timeout

	client, _ := vault.NewClient(config)
       
	ctx := context.Background()
	s, err := client.Auth().CreateWithContext(ctx, opts) // the context will never be cancelled here
```

Reply:
I think I get your idea. You want to wrap the context into a context with a timeout inside each `_WithContext` method?

Reply:
Like the method

```go
func (c *Client) wrapWithTimeout(ctx context.Context) (context.Context, context.CancelFunc) {
	if c.config.Timeout > 0 {
		return context.WithTimeout(ctx, c.config.Timeout)
	}
	return ctx, func() {}
}
```

And then call it like this
```go
func (c *TokenAuth) CreateWithContext(ctx context.Context, opts *TokenCreateRequest) (*Secret, error) {
	ctx, cancelFunc := c.c.wrapWithTimeout(ctx)
	defer cancelFunc()
```

Reply:
Yes, I think that would work 👍 

I would maybe call this helper function `withConfiguredTimeout` to be more explicit about its purpose. It would also follow the `WithXYZ` standard pattern.

Reply:
Done.

I have an idea. Since in all API methods that use context we call the `withConfiguredTimeout` method, we can simplify methods as follows:

```go
func (c *Client) withConfiguredTimeout(ctx context.Context) (context.Context, context.CancelFunc) {
	if c.config.Timeout > 0 {
		return context.WithTimeout(ctx, c.config.Timeout)
	}
	return context.WithCancel(ctx)
}

func (c *Client) Help(path string) (*Help, error) {
	return c.HelpWithContext(context.Background(), path)
}

// HelpWithContext reads the help information for the given path.
func (c *Client) HelpWithContext(ctx context.Context, path string) (*Help, error) {
	ctx, cancelFunc := c.withConfiguredTimeout(ctx)
	defer cancelFunc()
	// ...
}
```

What do you think?

Reply:
I don't it's necessary to change `withConfiguredTimeout`. Now that the context with timeout is always cancelled, we can simply remove the redundant `WithCancel(ctx)` calls:

```go
// unchanged
func (c *Client) withConfiguredTimeout(ctx context.Context) (context.Context, context.CancelFunc) {
	if c.config.Timeout > 0 {
		return context.WithTimeout(ctx, c.config.Timeout)
	}
	return ctx, func() {}
}

// removed context.WithCancel
func (c *Client) Help(path string) (*Help, error) {
	return c.HelpWithContext(context.Background(), path)
}
```

Reply:
Perhaps, yes. I'll change that.

Reply:
Done

Reply:
Looks great so far 👍 

One last thing here, the `RawRequestWithContext` function is invoked in [several places](https://github.com/hashicorp/vault/blob/7e74bebea84b098d9d2de0eed2490f352481717a/vault/raft.go#L1135) outside of the vault/api. Therefore we still need to keep the timeout hack inside of it. One way to solve this is to refactor the `RawRequestWithContext` implementation into unexported `rawRequestWithContext` (which will be called by all other vault/api functions) and add back the timeout hack (I have an example in my first comment).

Somewhat related to this, the `RawRequest` func should probably call the unexported `rawRequestWithContext` with a `withConfiguredTimeout(context.Background())` context.

Reply:
I replied inline with the change since this thread is getting too large.